### PR TITLE
Added readability-container-data-pointer clang-tidy check

### PR DESCRIPTION
**Pull request by gnawme**
Added the [readability-container-data-pointer](https://clang.llvm.org/extra/clang-tidy/checks/readability/container-data-pointer.html) check to `clang-tidy` to change instances where the code could use `data()` rather than the address of the element at index 0 in a container.

### CODE REVIEW

File: filters/src/passthrough.cpp [hunk 1]
```diff
@@ -217,7 +217,7 @@ pcl::PassThrough<pcl::PCLPointCloud2>::applyFilter (PCLPointCloud2 &output)
         }
 
         // Unoptimized memcpys: assume fields x, y, z are in random order
-        memcpy (&pt[0], &input_->data[xyz_offset[0]], sizeof(float));
+        memcpy (pt.data(), &input_->data[xyz_offset[0]], sizeof(float));
```
Comment (mvieth) [inline filters/src/passthrough.cpp:220]: Having `pt.data()` for the first one and `&pt[1/2]` for the others is kind of strange. Can you also check the other files for similar occurrences, please?

**Comment by gnawme** [comment]
Tests are passing on my desktop

File: recognition/src/face_detection/face_detector_data_provider.cpp [hunk 2]
```diff
@@ -282,15 +282,15 @@ void pcl::face_detection::FaceDetectorDataProvider<FeatureType, DataSet, LabelTy
     if (USE_NORMALS_)
     {
       integral_image_normal_x.reset (new pcl::IntegralImage2D<float, 1> (false));
-      const float *data_nx = reinterpret_cast<const float*> (&(*normals)[0]);
+      const float *data_nx = reinterpret_cast<const float*> (&(*normals)[0]);// NOLINT(readability-container-data-pointer)
```
Comment (larshg) [inline recognition/src/face_detection/face_detector_data_provider.cpp:285]: I guess one could just:
`const float* data = reinterpret_cast<const float*> (*normals).data();`

And in the two other, with y and z, have the call
`setInput(data + 1, ....)  `- for Y
and
`setInput(data + 2, ...)` - for Z

I'm not entirely sure what should be preferred, but currently all 3 statements return the same pointer, but calls them data_nx, data_ny and data_nz - which seems confusing to me 😄

File: recognition/src/face_detection/rf_face_detector_trainer.cpp [hunk 3]
```diff
@@ -302,15 +302,15 @@ void pcl::RFFaceDetectorTrainer::detectFaces()
   if (use_normals_)
   {
     integral_image_normal_x.reset (new pcl::IntegralImage2D<float, 1> (false));
-    const float *data_nx = reinterpret_cast<const float*> (&(*normals)[0]);
+    const float *data_nx = reinterpret_cast<const float*> (&(*normals)[0]);// NOLINT(readability-container-data-pointer)
```
Comment (larshg) [inline recognition/src/face_detection/rf_face_detector_trainer.cpp:305]: Same here.

File: recognition/src/face_detection/rf_face_detector_trainer.cpp [hunk 4]
```diff
@@ -276,7 +276,7 @@ void pcl::RFFaceDetectorTrainer::detectFaces()
 
   int element_stride = sizeof(pcl::PointXYZ) / sizeof(float);
   int row_stride = element_stride * cloud->width;
-  const float *data = reinterpret_cast<const float*> (&(*cloud)[0]);
+  const float *data = reinterpret_cast<const float*> (&(*cloud)[0]); // NOLINT(readability-container-data-pointer)
```
Comment (larshg) [inline recognition/src/face_detection/rf_face_detector_trainer.cpp:279]: Why not using .data() here?

**Comment by larshg** [comment]
I think the approach with using .data() to get the first index and then add +1, +2 on subsequent usages, for ie. X,Y,Z values seems more readable, than using the &ofsomething[0/1/2] - what you think @gnawme and @mvieth ?

**Comment by gnawme** [comment]
> I think the approach with using .data() to get the first index and then add +1, +2, +X on subsequent usages, for ie. X,Y,Z values seems more readable, than using the &ofsomething[0/1/2] - what you think @gnawme and @mvieth ?

Makes sense to me, I hope my changes reflect your suggestion properly.

**Comment by mvieth** [comment]
> I think the approach with using .data() to get the first index and then add +1, +2, +X on subsequent usages, for ie. X,Y,Z values seems more readable, than using the &ofsomething[0/1/2] - what you think @gnawme and @mvieth ?

I am fine with either, as long as it is used consistently

**Review by mvieth — APPROVED** [review]

File: recognition/src/face_detection/face_detector_data_provider.cpp [hunk 5]
```diff
@@ -254,7 +254,7 @@ void pcl::face_detection::FaceDetectorDataProvider<FeatureType, DataSet, LabelTy
 
     int element_stride = sizeof(pcl::PointXYZ) / sizeof(float);
     int row_stride = element_stride * cloud->width;
-    const float *data = reinterpret_cast<const float*> (&(*cloud)[0]);
+    const float *data = reinterpret_cast<const float*> (&(*cloud)[0]); // NOLINT(readability-container-data-pointer)
```
Comment (larshg) [inline recognition/src/face_detection/face_detector_data_provider.cpp:257]: Why not .data() here?

**Review by larshg — APPROVED** [review]

**Pull request merged**